x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java

```diff
@@ -0,0 +1,298 @@
+        final ShardId shardId,
+        final String expectedHistoryUUID,
+        long fromSeqNo,
+        final ByteSizeValue maxBatchSize,
+        final TimeValue pollTimeout,
+        int maxOperationsCount
+    ) {
+        final ShardChangesAction.Request shardChangesRequest = new ShardChangesAction.Request(
+            new ShardId(new Index(indexName, indexUUID), shardId.id()),
+            expectedHistoryUUID
+        );
+        shardChangesRequest.setFromSeqNo(fromSeqNo);
+        shardChangesRequest.setMaxBatchSize(maxBatchSize);
+        shardChangesRequest.setPollTimeout(pollTimeout);
+        shardChangesRequest.setMaxOperationCount(maxOperationsCount);
+        return shardChangesRequest;
+    }
+
+    /**
+     * Converts the response to XContent JSOn format.
+     *
+     * @param response The ShardChangesAction response.
+     * @param indexName The name of the index.
+     * @param shardId The ShardId.
+     */
+    private static XContentBuilder shardChangesResponseToXContent(
+        final ShardChangesAction.Response response,
+        final String indexName,
+        final ShardId shardId
+    ) {
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            builder.startObject();
+            builder.field("index", indexName);
+            builder.field("shard_id", shardId);
+            builder.field("mapping_version", response.getMappingVersion());
+            builder.field("settings_version", response.getSettingsVersion());
+            builder.field("aliases_version", response.getAliasesVersion());
+            builder.field("global_checkpoint", response.getGlobalCheckpoint());
+            builder.field("max_seq_no", response.getMaxSeqNo());
+            builder.field("max_seq_no_of_updates_or_deletes", response.getMaxSeqNoOfUpdatesOrDeletes());
+            builder.field("took_in_millis", response.getTookInMillis());
+            if (response.getOperations() != null && response.getOperations().length > 0) {
+                operationsToXContent(response, builder);
+            }
+            builder.endObject();
+
+            return builder;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Converts the operations from a ShardChangesAction response to XContent JSON format.
+     *
+     * @param response The ShardChangesAction response containing the operations to be converted.
+     * @param builder The XContentBuilder to which the converted operations will be added.
+     * @throws IOException If an error occurs while writing to the XContentBuilder.
+     */
+    private static void operationsToXContent(final ShardChangesAction.Response response, final XContentBuilder builder) throws IOException {
+        builder.field("number_of_operations", response.getOperations().length);
+        builder.field("operations");
+        builder.startArray();
+        for (final Translog.Operation operation : response.getOperations()) {
```

Comment:
Maybe we can avoid this...as serializing each operation for large requests might be way too much?

Reply:
I see that you already left out the actual content of the write operations, which is good for the use case this rest api is going to be used. Let's for now keep the response as is here. If this turns out the be problematic we can always drop the operations array from the response.